src/Hazelcast.Net/CP/CpSubsystemSession.cs

```diff
@@ -0,0 +1,264 @@
+        /// </summary>
+        /// <param name="groupId"></param>
+        /// <returns>Session id or <see cref="NoSessionId"/> if absent</returns>
+        public long GetSessionId(CPGroupId groupId)
+        {
+            if (_sessions.TryGetValue(groupId, out var sessionState))
+                return sessionState.Id;
+            else
+                return NoSessionId;
+        }
+
+        public async Task CloseSessionAsync(CPGroupId groupId, long sessionId)
+        {
+            InvalidateSession(groupId, sessionId);
+            await RequestCloseSessionAsync(groupId, sessionId).CfAwait();
+        }
+
+        /// <summary>
+        /// Shuts down sessions on server and disposes
+        /// </summary>
+        /// <returns></returns>
+        public async Task ShutdownAsync()
+        {
+            await _semaphoreReadWrite.WaitAsync().CfAwait();
+
+            try
+            {
+                await _sessions.Keys
+                    .ParallelForEachAsync(async (key, cancelToken) =>
+                        {
+                            try
+                            {
+                                await CloseSessionAsync(key, _sessions[key].Id).CfAwait();
+                            }
+                            catch (Exception e)
+                            {
+                                _logger.LogWarning("Exception thrown while closing CP sessions", e);
+                            }
+                        },
+                        _cancel.Token)
+                    .CfAwait();
+
+                await DisposeAsync().CfAwait();
+            }
+            finally { _semaphoreReadWrite.Release(); }
+        }
+
+        /// <summary>
+        /// Gets or createas a session if absent by group id.
+        /// </summary>
+        /// <param name="groupId"></param>
+        /// <returns><see cref="CPSubsystemSessionState"/></returns>
+        /// <exception cref="HazelcastInstanceNotActiveException"></exception>
+        private async Task<CPSubsystemSessionState> GetOrCreateSessionAsync(CPGroupId groupId)
+        {
+            await _semaphoreReadWrite.WaitAsync().CfAwait();
+
+            try
+            {
+                if (_disposed == 1) throw new ObjectDisposedException("CP Subsystem Session is already disposed!");
+
+                if (_sessions.TryGetValue(groupId, out var sessionState) && sessionState.IsValid)
+                {
+                    return sessionState;
+                }
+                else
+                {
+                    // Wait and lock only for the groupId
+                    var semaphore = GetSemaphoreBy(groupId);
+                    await semaphore.WaitAsync().CfAwait();
+
+                    try
+                    {
+                        var session = await RequestNewSessionAsync(groupId).CfAwait();
+                        ScheduleHeartbeat(TimeSpan.FromMilliseconds(session.Item2));
+                        _sessions[groupId] = session.Item1;
```

Comment:
java uses this order:

```suggestion
                        _sessions[groupId] = session.Item1;
                        ScheduleHeartbeat(TimeSpan.FromMilliseconds(session.Item2));
```
puts the session in the map even if the schedule heartbeat throws.

Reply:
Fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/590/commits/993cabbcb9bac708cc38ea83239b98cf19ead875